### PR TITLE
fix: settle session-authenticated chat instead of serving it free (#746, #856)

### LINKED ISSUE #856: Console overview and analytics read zero after live chat traffic on the same workspace
## What happens

Two chat completions were served successfully to the demo account through `hive-auto` on 2026-08-11. Minutes later, the console for the same workspace reports nothing:

* Overview: Available credits `0`, Posted `0`, Reserved `0`, Today's activity renders as an em dash with the caption "Detailed counts available in Analytics", Recent errors renders as an em dash with the caption "Error telemetry not yet wired up".
* Analytics, 24h window: TOTAL REQUESTS `0`, INPUT TOKENS `0`, OUTPUT TOKENS `0`, CREDITS SPENT `0`.

## Why it matters for a demo

The console is the part of the product that proves the gateway is a metered, governed platform rather than a chat window. Walking an audience from a live conversation to a usage page that reads zero invites the obvious question, and there is no good answer to it on stage.

There is a second, larger question hiding behind the same screens. The workspace balance is zero and requests were still served. Either the chat surface is not attributed to this workspace at all, or the requests were served without being metered. Both readings deserve a look, and the second one is a billing correctness matter, not a presentation matter.

## Evidence

Overview after live chat traffic:

Console overview reads zero

Analytics after live chat traffic, 24h window:

Analytics reads zero

Proof the traffic was real, same account, same day:

Model answered

## Reproduction

1. Sign in to chat as the demo account and send a prompt that receives an answer.
2. Open `https://console-hive.scubed.co/console` and `https://console-hive.scubed.co/console/analytics` for the same workspace.
3. Every counter reads zero, in every time window.

## Scope

Two separable questions, worth splitting once someone has looked:

1. Presentation. Is chat traffic supposed to appear in this workspace's analytics, and if it is, where does the attribution break.
2. Metering. A zero balance served two completions. Confirm whether a hold was placed and settled, or whether the request was served free.

Verified live on 2026-08-11 against the deployed stack.

### PR DESCRIPTION
**Pull request by sakibsadmanshajib**
Closes #746. Closes #856.

## What was wrong

Session-authenticated chat served inference and recorded nothing. `jwtAwareChatHandler` (`apps/edge-api/cmd/server/main.go`) sends a request with a populated `auth.UserFrom` to `chat.Handler`, which wrote `llm_traces` and `audit_log` and nothing else: no hold, no settlement, no verdict. The budget middleware could not cover it either, because it resolves workspace identity from an `hk_` bearer alone, so a workspace showing a zero balance still served completions. `rewriteDispatchBody` never set `stream_options.include_usage`, so the terminal usage frame never arrived and the token counts behind any charge were always zero.

The branch dates to 2026-05-18 and was dormant until PR #735 gave Open WebUI a per-user token, which moved every chat turn onto it. Eighty traces across four tenants were served that way, each recording zero tokens and zero credits. #856 is the same hole seen from the console: analytics reads `usage_events`, the session path wrote none, so every counter read zero for a workspace that had just held a conversation.

## The fix

`apps/edge-api/internal/chat/billing.go` is the new accounting half of the dispatch path.

Before the provider is reached:

1. The tenant's billing account and deployment posture resolve in one indexed lookup (`metering.PGBillingAccountResolver.ResolveState`).
2. An `ENTERPRISE_EDGE` tenant is explicitly not billable per D-027. It takes no hold, and it is checked before the pricing gate so an Enterprise box running a locally hosted unpriced model keeps working.
3. An alias the catalog cannot price in tokens is refused rather than recorded as an unpriced trace (D-034).
4. A tenant with no `tenant_billing_accounts` row is refused with `billing_not_configured` rather than served free.
5. A 10000 credit hold is taken, the same figure the API-key chat path reserves. A control-plane credit rejection (409) or a rate limit (429) becomes an `insufficient_quota` refusal, classified on status codes and never on message text. Unlike the API-key path, every reservation failure refuses here, including a control-plane fault: route selection is itself a control-plane call one step earlier, so an outage has already refused the request before this line is reached.

During dispatch, the body carries `stream_options.include_usage` through `metering.RewriteBody`, so the token counts are measured rather than assumed.

After the stream ends, the reservation reaches a terminal state exactly once: finalized at the catalog price of the route that served the request, or released in full. A failed finalize falls back to a release, so a lost charge cannot also strand a hold (#616). `FinalizeReservation` now carries the metered token counts and control-plane writes them onto the settlement usage event, so one row states both what was consumed and what it cost, which is what makes the console counters read a real figure rather than zero.

## What I broke to see the test fail

`TestSessionChatSettlesAtTheCatalogMagnitude` asserts the magnitude of the charge, not the existence of a row. To confirm it can actually go red, the settled figure in `dispatch.go` was scaled by a constant:

```go
case settle.finalize(costCredits*2, confirmed, int64(inTokens), int64(outTokens)):
```

Both table cases failed on magnitude, and only on magnitude:

```
--- FAIL: TestSessionChatSettlesAtTheCatalogMagnitude/hive-auto
    Error:      Not equal:
                expected: 22
                actual  : 44
    Messages:   settled credits must equal the catalog price of the metered tokens
--- FAIL: TestSessionChatSettlesAtTheCatalogMagnitude/hive-default
    Error:      Not equal:
                expected: 5
                actual  : 10
```

The multiplier was then removed and the suite is green again. The expected figures come from a second implementation of the arithmetic written inside the test (`expectedCredits`) and are cross-checked against hand-computed literals, so the assertion cannot pass by calling the production helper and comparing it with itself:

* `hive-auto` at 56000 in and 224000 out per million, 111 in and 70 out tokens, `(111*56000 + 70*224000) / 1e6 = 21.896`, settles at 22.
* `hive-default` at 21000 in and 84000 out per million, 111 in and 32 out tokens, `(111*21000 + 32*84000) / 1e6 = 5.019`, settles at 5.

`TestFinalizeReservationTwiceChargesOnce` was also verified non-vacuous: disabling the replay guard in `finalizeLocked` made it fail with `second FinalizeReservation returned error: finalized reservations cannot be finalized`, and the guard was restored.

## Invariant coverage, by test rather than by inspection

| Invariant | Test |
| --- | --- |
| Exactly one settled entry per served completion, never zero, never two | `TestSessionChatSettlesAtTheCatalogMagnitude` asserts one reservation, one finalize and zero releases. `TestFinalizeReservationTwiceChargesOnce` (control-plane) asserts one charge and one settlement event across two finalize calls |
| Attributed to the tenant and workspace that made the request | The same test asserts the resolved account id on both the reservation and the settlement, and the requested alias on the reservation |
| Priced at the route that actually served it | Two aliases at different catalog prices settle at two different magnitudes, both derived from the route selection result rather than a constant. The alias is never the price source. A served model that differs from the dispatched route is detected and logged for #743 |
| No charge on a refused or failed request, no stranded hold | `TestSessionChatRefusesWhenTheAccountCannotPay` and `TestSessionChatRefusesTenantWithNoBillingAccount` assert zero upstream calls and zero settlements. `TestSessionChatReleasesTheHoldWhenTheUpstreamFails` and `TestSessionChatReleasesTheHoldWhenSettlementFails` assert exactly one release and zero charges |
| Token counts are real, not estimated | `TestSessionChatRequestsTheUsageEnvelope` asserts `stream_options.include_usage` on the outbound body, and the magnitude test asserts the provider's counts reach the settlement with `terminal_usage_confirmed` true |

`TestSessionChatEnterpriseTenantIsNotBilled` records the one deliberate exception, and `TestSessionChatRefusesWhenAccountingIsNotWired` asserts that a gateway built without its accounting seam refuses rather than repeating the defect.

## Deliberate non-goals

* The 80 unbilled requests are not retro-priced. The token counts were never captured, so the data does not exist and inventing it would be worse than the gap.
* The `hive_credit_delta` sign and unit disagreement on the API-key path is untouched. It has its own issue, and rewriting financial history is an owner decision.
* No unique index on `usage_events (request_id, event_type)` is added. Today the API-key path writes two rows with the same pair for one request (control-plane's settlement plus edge-api's completed telemetry), so such an index would fail to create and would break that path. Exactly-once is enforced here by a single settlement site per request plus control-plane's replay guard, and the index question belongs with the duplicate-row defect.
* `metering.Gate` stays unwired. Its `Execute` dispatches first and grades afterwards, which is the wrong shape for a refusal that has to happen before the provider call, and a shadow verdict alongside a real settlement would be double bookkeeping.

## The routing lint, and why it moved

`lint:litellm-routing` went red on this branch. It reported that
`apps/edge-api/internal/metering` is handed the LiteLLM address or master key
in `cmd/server/main.go` and never calls `SelectRoute`. It is not waived and it
has no allowlist entry: this pull request exists because a dispatch path
skipped metering, and silencing that guard inside the fix would be the worst
available resolution.

`metering` is not handed either credential. It receives `Pool` and nothing
else. The lint's attribution rule was wrong, and wrong in the direction that
matters. It walked back from the resolver call to the nearest preceding
package-qualified identifier on any line, so adding

```go
Billing: &metering.PGBillingAccountResolver{Pool: dbPool},
```

above `LiteLLMURL:` inside the same `chat.Deps` literal moved attribution onto
`metering`. That is a false positive on a package that never receives the
credential and, more seriously, a false negative on `chat`, the package that
does, which stopped being checked by this rule at all. Demonstrated on the old
rule directly:

```go
rogue := limits.NewBudgetGate(limits.Deps{
    Metrics:     metrics.NewCounter(),
    UpstreamURL: resolveLiteLLMBaseURL(),
})
```

The old rule attributes that to `metrics` and never names `limits`. A guard
that can be pointed away from any real recipient by inserting one
package-qualified sibling field is not enforcing the thing it exists to
enforce, and the wiring rule's entire purpose is catching a dispatcher whose
own identifiers avoid the word "litellm".

Attribution is now to the type of the composite literal or call that lexically
encloses the resolver, located with a brace and paren counter that carries the
opener's line and column, then reading the qualified type immediately to the
left of that opener. The column matters: on a single-line literal the last
qualifier anywhere on the line is a sibling's expression rather than the
literal's own type.

The self-test covers the sibling-field, nested-literal, single-line and
positional-argument shapes alongside the original rogue case, and asserts both
halves of each (the real recipient is named AND the sibling is not, since a
rule that attributed both would pass a one-sided assertion). Checked against
the real tree as well as the self-test: wiring either rogue shape above into
`main.go` makes the lint exit non-zero naming `apps/edge-api/internal/limits`,
and removing it returns it to green. The rule is strictly stronger than the one
it replaces, not relaxed.

## Review threads

* `writeUnpriceableModel` answered 503 `api_error` for a permanent condition,
  telling every SDK retry layer to resend a request that can never succeed. Now
  400 `invalid_request_error`, with wording that asks for a different model
  rather than a retry.
* `PGBillingAccountResolver.ResolveState` treated a nil pool as a successful
  lookup that found nothing, so a database outage was reported as the permanent
  403 `billing_not_configured` rather than the retryable
  `billing_unavailable`, and the accounting-not-wired guard could not fire
  (a nil `*PGBillingAccountResolver` in an interface field is not a nil
  interface). It now returns `metering.ErrNoPool`, at the resolver so every
  caller including `Gate` is corrected at once.
* A finalize whose answer is lost is retried once before falling back to a
  release. The charge may already have committed, and reading that as uncharged
  releases a hold the ledger has settled and records `charged=false` against a
  customer who was charged. It cannot double charge: `finalizeLocked` replays
  the stored row for an already settled reservation.
* `delta.refusal` is accumulated alongside `delta.content`. A refusal is
  delivered output that cost provider tokens, and counting only content made a
  refusal-only answer with no usage frame look undelivered, which released the
  hold and served it free. The API-key accumulator has always counted it.
* Provider token counts are clamped at the finalize boundary, so a negative
  count cannot reach `usage_events` and sit in console analytics beside a
  non-negative credit delta.

Every new assertion was checked non-vacuous by reverting its fix and watching
it fail. `TestSessionChatReleasesTheHoldWhenSettlementFails` now asserts two
finalize attempts and still exactly one release, which is a stronger statement
than the single attempt it asserted before.

## Deployment note

Any tenant without a `tenant_billing_accounts` row now receives a refusal on
session chat instead of free inference. This is the intended fail-closed
behaviour under D-034; provisioning coverage is tracked in #721.

The E2E fixtures were the reachable case. `e2e-verified-tenant` had no billing
account and was producing traces, so after this change it would have been
refused and the suite would have gone red in a way that looked like this fix
broke it. The refusal is not weakened. Instead `seedBilling` in
`tests/e2e/support/e2e-fixture-seed.mjs` maps each fixture tenant to the
account it settles against and grants it a balance several holds deep, so the
suite exercises the real metered path rather than the refusal in front of it. A
test tenant that cannot be billed is not a realistic test of a billing gateway.

That needed one change to the stale-run sweep. `tenant_billing_accounts` pins
`account_id` with `ON DELETE RESTRICT` so deleting an account that still funds
a live tenant fails loudly, so the sweep now unmaps before it deletes;
otherwise the account delete would fail the foreign key on every stale run and
the sweep would quietly stop reclaiming anything.

Existing rows were backfilled in the same shape: 462 of the 477 tenants are E2E
fixture leftovers, and every one whose account still exists is now mapped and
funded, which took mapped tenants from 26 to 52. The remaining 422 have had
their accounts swept already, so there is no account left to bill and no user
left to sign in as them.

## Verification

```
go test ./apps/edge-api/... -count=1 -short      # pass
go test ./apps/control-plane/... -count=1 -short # pass
go build ./apps/edge-api/... ./apps/control-plane/...
go vet ./apps/edge-api/internal/chat/... ./apps/edge-api/internal/metering/...
```

All run in the repo toolchain container. The new tests need no database, so they run in the required `Go tests (edge-api)` step rather than only in the DB-backed one.

No UI surface changes in this PR, so no screenshot proof applies. The console reading real figures is a consequence of the writer fix and is worth capturing live once this is deployed to the demo box.

## Summary by CodeRabbit

- **New Features**
  - Session-authenticated chat now supports credit reservation and usage-based billing.
  - Charges reflect measured input and output token usage.
  - Enterprise deployments remain exempt from billing.
  - Refusal-only responses are settled correctly.

- **Bug Fixes**
  - Prevented chat requests from bypassing billing and usage recording.
  - Added safeguards for missing billing accounts, insufficient credits, and unsupported pricing.
  - Reservations are released when upstream requests or settlement fail.
  - Settlement retries safely before releasing holds.
  - Duplicate settlement no longer creates duplicate charges.

### CODE REVIEW

File: apps/edge-api/internal/chat/billing.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+	code := "reservation_unavailable"
+	apierrors.WriteError(w, http.StatusServiceUnavailable, "api_error",
+		"Credit reservation is temporarily unavailable. Please retry.", &code)
+}
+
+func writeUnpriceableModel(w http.ResponseWriter) {
+	code := "model_not_supported"
+	apierrors.WriteError(w, http.StatusServiceUnavailable, "api_error",
+		"The requested model is not available for this endpoint. Please retry with a supported model.", &code)
+}
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/edge-api/internal/chat/billing.go:292]: Fixed in 38cb0a1a. Agreed, and the retry-storm argument is the sharper half of it: an alias with no token price is a stable property of its catalog row, so a 503 with type api_error told every SDK retry layer to resend a request that can never succeed.

It now answers 400 with type invalid_request_error, and the wording asks for a different model rather than a retry, so the status and the message finally say the same thing.

Test added as asked: TestSessionChatRefusesAnUnpriceableAliasPermanently asserts the 4xx, asserts the status is below 500 so a future change cannot quietly reclassify it as a transient fault, and asserts zero reservations, zero finalizes, zero releases and zero upstream calls. It was checked non-vacuous by restoring the 503 and watching it fail.

File: apps/edge-api/internal/metering/precedence.go [hunk 2]
```diff
@@ -65,19 +65,40 @@ type PGBillingAccountResolver struct {
...
+// ResolveState answers both questions the money path asks about a tenant in
+// one indexed lookup: which account its usage settles against, and the
+// deployment posture that decides whether it settles at all. A tenant row with
+// no tenant_billing_accounts match returns Found false, which the enforcing
+// caller (apps/edge-api/internal/chat) refuses on rather than serving free.
+func (r *PGBillingAccountResolver) ResolveState(ctx context.Context, tenantID uuid.UUID) (TenantBillingState, error) {
 	if r == nil || r.Pool == nil {
-		return uuid.Nil, false, nil
+		return TenantBillingState{}, nil
 	}
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/edge-api/internal/metering/precedence.go:92]: Fixed in 38cb0a1a. Real defect, and the consequence is worse than the status code alone suggests: an unknown billing position and a tenant known to have no billing account are different answers, and collapsing them reports a database outage as a workspace-provisioning problem that no client will retry and an operator will chase in the wrong place.

Taken at the resolver rather than at the wiring site. ResolveState now returns a new sentinel, metering.ErrNoPool, when it has no pool. That is the root-cause position: every caller routes through it, including Gate.resolveBillingAccount, which already maps an error to billing_unavailable, so both consumers get the correct classification from one change.

The suggested main.go half was deliberately not taken. Deps.Billing is an interface, so assigning a nil *PGBillingAccountResolver to it produces a non-nil interface value, and the accounting-not-wired guard in startSettlement would still not fire. A conditional at the wiring site would work but only for that one call site, whereas the sentinel covers any future one.

TestSessionChatReportsAnUnknownBillingPositionAsRetryable uses the real production resolver with no pool, exactly the shape cmd/server wires when SUPABASE_DB_URL is missing, and asserts 503 billing_unavailable, the absence of billing_not_configured, and that no inference was served.

File: apps/edge-api/internal/metering/precedence.go [hunk 3]
```diff
@@ -65,19 +65,40 @@ type PGBillingAccountResolver struct {
...
+
+// ResolveState answers both questions the money path asks about a tenant in
+// one indexed lookup: which account its usage settles against, and the
+// deployment posture that decides whether it settles at all. A tenant row with
+// no tenant_billing_accounts match returns Found false, which the enforcing
+// caller (apps/edge-api/internal/chat) refuses on rather than serving free.
+func (r *PGBillingAccountResolver) ResolveState(ctx context.Context, tenantID uuid.UUID) (TenantBillingState, error) {
 	if r == nil || r.Pool == nil {
-		return uuid.Nil, false, nil
+		return TenantBillingState{}, nil
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/edge-api/internal/metering/precedence.go]: Fixed in 38cb0a1a, same defect as the CodeRabbit thread on this file, taken at the resolver.

ResolveState returns a new sentinel, metering.ErrNoPool, when it has no pool, so an unknown billing position stops looking like a tenant known to have no billing account. startSettlement already routes an error to the retryable billing_unavailable, and Gate.resolveBillingAccount already maps one to the same code, so one change at the shared function corrects both callers rather than only the call site named here.

The alternative you offered, not wiring the resolver when the pool is nil, was not taken on its own: Deps.Billing is an interface, so a nil *PGBillingAccountResolver assigned to it is still a non-nil interface value and the accounting-not-wired guard would not fire. The sentinel does not depend on the wiring site getting that subtlety right.

TestSessionChatReportsAnUnknownBillingPositionAsRetryable covers it with the real resolver and no pool.

File: apps/edge-api/internal/chat/billing.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+		// full past the hold and skips reconciliation.
+		TerminalUsageConfirmed: confirmed,
+		InputTokens:            inTokens,
+		OutputTokens:           outTokens,
+		Status:                 "completed",
+	})
+	if err != nil {
+		slog.Error("session chat finalize failed, releasing hold instead",
+			"err", err, "request_id", s.requestID, "reservation_id", s.reservationID)
+		return false
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/edge-api/internal/chat/billing.go]: Fixed in 38cb0a1a. This was the sharpest of the five: a lost answer is not the same evidence as a refusal, and treating it as uncharged releases a hold the ledger has already settled while writing charged=false and cost_credits=0 against a customer who was in fact charged.

finalize now retries once on its own fresh context before falling back to a release. Re-reading reservation state was the other option you offered and it is strictly more work for the same answer, because FinalizeReservation is already idempotent by construction: finalizeLocked replays the stored row for a reservation already in finalized or needs_reconciliation rather than charging again. So the retry either confirms the charge that committed or lands the one that never did, and it cannot produce a second charge. TestFinalizeReservationTwiceChargesOnce is what holds that property in place, and the retry is now a second caller depending on it.

Two tests. TestSessionChatRetriesAnAmbiguousFinalizeBeforeReleasing fails the first finalize with a 504 and asserts exactly two attempts, zero releases, and that both attempts carry the identical charge and token counts, so a retry cannot settle at a different magnitude than the attempt whose answer was lost. TestSessionChatReleasesTheHoldWhenSettlementFails now asserts two finalize attempts and still exactly one release, which is a stronger statement than the single attempt it asserted before, not a relaxed one.

File: apps/edge-api/internal/chat/dispatch.go [hunk 5]
```diff
@@ -230,21 +265,54 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	// The charge is the catalog price of the route that actually served the
+	// request applied to the tokens the provider reported, the same conversion
+	// the API-key path settles with (#688), so the two surfaces cannot report
+	// different costs for identical usage. When no usage frame arrives at all,
+	// the same helper falls back to a content estimate and flags it
+	// unconfirmed, which is what tells control-plane to clamp the figure to the
+	// hold and open a reconciliation job. It never settles a delivered response
+	// at zero, and never bills a token count as though it were a credit amount.
+	costCredits, confirmed, delivered := inference.ChatSettlementCredits(
+		route, hasUsage, int64(inTokens), int64(outTokens), raw, completion.String())
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/edge-api/internal/chat/dispatch.go:289]: Fixed in 38cb0a1a. Correct, and it is precisely the failure mode this pull request exists to remove: a customer receives output and is charged nothing for it.

sseEnvelope now carries delta.refusal and the stream loop appends it to the same accumulator as delta.content, matching inference.UsageAccumulator.Add on the API-key path. A refusal is delivered output that cost provider tokens to produce, so it belongs in the delivered-content signal and in the fallback estimate alike.

TestSessionChatSettlesARefusalOnlyAnswer streams a refusal-only answer with no terminal usage frame, which is the only shape where the accumulated text decides whether anything was delivered, and asserts one finalize, zero releases, a positive charge, and TerminalUsageConfirmed false so the estimate is flagged for reconciliation rather than billed as measured truth. Reverting the two-line accumulation makes it fail.

File: apps/control-plane/internal/accounting/service.go [hunk 6]
```diff
@@ -363,8 +363,13 @@ func (s *Service) finalizeLocked(ctx context.Context, input FinalizeReservationI
 		Endpoint:         reservation.Endpoint,
 		ModelAlias:       reservation.ModelAlias,
 		Status:           string(status),
-		HiveCreditDelta:  -actualCredits,
-		CustomerTags:     reservation.CustomerTags,
+		// The quantities behind the charge, when the caller measured them, so
+		// this one row answers both what was consumed and what it cost. Spend
+		// stays the negative credit delta; these are counts, never credits.
+		InputTokens:     input.InputTokens,
+		OutputTokens:    input.OutputTokens,
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/control-plane/internal/accounting/service.go]: Fixed in 38cb0a1a. Agreed. CreditsForTokens clamps before pricing, so the charge was never at risk, but the settlement usage event is read by console analytics and a negative count would sit there beside a non-negative credit delta, where a SUM over the column silently understates consumption. These counts come from a provider response, so they are external input on a money path and belong clamped at the boundary that records them.

finalizeLocked now clamps both to zero before building the RecordEventInput, which is the finalize boundary you named and the single chokepoint every settlement path routes through.

TestFinalizeClampsNegativeProviderTokenCounts finalizes with -111 and -70 and asserts the settlement event records zero and zero with the credit delta still -22. Removing the clamp makes it fail with the negative counts printed.

File: tools/lint-litellm-dispatch-requires-routing.mjs [hunk 7]
```diff
@@ -89,20 +89,52 @@ function dispatchesToLiteLLM(source) {
...
+    // type. `Billing: &metering.PGBillingAccountResolver{...}` sitting above the
+    // credential is a sibling field, never the recipient; the recipient is the
+    // literal both fields belong to.
+    for (let at = index; at >= 0; ) {
+      const opener = enclosingConstructorLine(lines, at);
+      if (opener < 0) return;
+      const matches = [...lines[opener].matchAll(PACKAGE_QUALIFIER)];
       if (matches.length > 0) {
         packages.add(matches[matches.length - 1][1]);
         return;
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply tools/lint-litellm-dispatch-requires-routing.mjs]: Correct, and it is the same defect one line further down. The walk found the enclosing opener's line but then read the last qualifier anywhere on that line, so the single-line form you wrote recorded telemetry and left rogue free to dispatch unrouted, which is exactly the false negative this change set out to remove.

Fixed in eb15a57d as you specified. enclosingOpener now returns the opener's line and column, and qualifierBefore selects the package-qualified type whose match ends at or before that column, so it reads the literal's own type rather than a sibling's expression. The outward walk carries the column too, so a nested literal on the same line behaves the same way.

Self-test case added for the single-line shape, asserting both halves: rogue is attributed and telemetry is not. Asserting only the first would pass on a rule that attributed both.

Also checked against the real tree rather than the self-test alone: wiring precisely

    rogue := limits.NewBudgetGate(limits.Deps{Metrics: metrics.NewCounter(), UpstreamURL: resolveLiteLLMBaseURL()})

into cmd/server/main.go now makes the lint exit non-zero naming apps/edge-api/internal/limits, where before this commit it passed.

File: apps/control-plane/internal/accounting/service_test.go [hunk 8]
```diff
@@ -864,3 +864,121 @@ func projectedBudgetWindow(calls []apiKeyDeltaCall, apiKeyID uuid.UUID, budgetKi
...
+	}
+	if len(settlements) != 1 {
+		t.Fatalf("expected exactly one settlement usage event, got %d", len(settlements))
+	}
+	if settlements[0].InputTokens != 0 || settlements[0].OutputTokens != 0 {
+		t.Fatalf("negative provider token counts reached usage_events: %+v", settlements[0])
+	}
+	if settlements[0].HiveCreditDelta != -22 {
+		t.Fatalf("expected a -22 credit delta, got %d", settlements[0].HiveCreditDelta)
+	}
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/control-plane/internal/accounting/service_test.go:992]: Fair, and it is the failure mode worth guarding: the usage event is written from input.ActualCredits regardless of whether the ledger was actually touched, so a regression that dropped ChargeUsage would leave the assertion reading -22 and the test green while nothing was charged.

Fixed in eb15a57d. The test now asserts exactly one ledger charge of twenty two credits before it looks at the event, matching what TestFinalizeReservationTwiceChargesOnce already asserts alongside it.
